File: mobile/src/types/lesson.ts
/**
 * 五种教学组件契约（与 docs/lesson-components.md 一一对应）。
 * mock 数据在此基础上附带 sectionId / sectionTitle 用于分节。
 */

export interface WordComponent {
  type: "word";
  word: string;
  meaning: string;
  example: string;
  exampleMeaning: string;
  message: string;
}

export interface SentenceComponent {
  type: "sentence";
  sentence: string;
  meaning: string;
  message: string;
}

export type RecallMode = "zh_to_en" | "en_to_zh" | "audio_to_text" | "fill_blank";

export interface RecallComponent {
  type: "recall";
  mode: RecallMode;
  prompt: string;
  answer: string;
  message: string;
}

export interface PronunciationComponent {
  type: "pronunciation";
  content: string;
  meaning: string;
  message: string;
}

export interface DialogComponent {
  type: "dialog";
  scene: string;
  goal: string;
  opening: string;
}

export type LessonComponent =
  | WordComponent
  | SentenceComponent
  | RecallComponent
  | PronunciationComponent
  | DialogComponent;

type SectionTag = { sectionId?: string; sectionTitle?: string };

/** mock 数据里的活动 = 契约组件 + 分节标记 */
export type Activity = LessonComponent & SectionTag;

/** 单节课（按 sectionId 分组后） */
export interface LessonSection {
  id: string;
  title: string;
  activities: Activity[];
}

File: server/app/schemas/lesson.py
"""教学组件服务端契约；字段与 docs/lesson-components.md 保持一致。"""
from typing import Annotated, Literal, Union

from pydantic import BaseModel, ConfigDict, Field


class StrictComponent(BaseModel):
    model_config = ConfigDict(extra="forbid")


class WordComponent(StrictComponent):
    type: Literal["word"]
    word: str = Field(min_length=1, max_length=120)
    meaning: str = Field(min_length=1, max_length=240)
    example: str = Field(min_length=1, max_length=500)
    exampleMeaning: str = Field(min_length=1, max_length=500)
    message: str = Field(min_length=1, max_length=500)


class SentenceComponent(StrictComponent):
    type: Literal["sentence"]
    sentence: str = Field(min_length=1, max_length=500)
    meaning: str = Field(min_length=1, max_length=500)
    message: str = Field(min_length=1, max_length=500)


class RecallComponent(StrictComponent):
    type: Literal["recall"]
    mode: Literal["zh_to_en", "en_to_zh", "audio_to_text", "fill_blank"]
    prompt: str = Field(min_length=1, max_length=500)
    answer: str = Field(min_length=1, max_length=500)
    message: str = Field(min_length=1, max_length=500)


class PronunciationComponent(StrictComponent):
    type: Literal["pronunciation"]
    content: str = Field(min_length=1, max_length=500)
    meaning: str = Field(min_length=1, max_length=500)
    message: str = Field(min_length=1, max_length=500)


class DialogComponent(StrictComponent):
    type: Literal["dialog"]
    scene: str = Field(min_length=1, max_length=200)
    goal: str = Field(min_length=1, max_length=800)
    opening: str = Field(min_length=1, max_length=500)


LessonComponent = Annotated[
    Union[WordComponent, SentenceComponent, RecallComponent, PronunciationComponent, DialogComponent],
    Field(discriminator="type"),
]


class LessonSection(BaseModel):
    model_config = ConfigDict(extra="forbid")
    id: str = Field(min_length=1, max_length=96, pattern=r"^[A-Za-z0-9_.-]+$")
    title: str = Field(min_length=1, max_length=255)
    activities: list[LessonComponent] = Field(min_length=1, max_length=500)


class CourseContent(BaseModel):
    model_config = ConfigDict(extra="forbid")
    title: str = Field(min_length=1, max_length=255)
    intro: str = Field(default="", max_length=2000)
    sections: list[LessonSection] = Field(min_length=1, max_length=100)
